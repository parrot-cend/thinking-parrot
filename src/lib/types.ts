export interface Dictionary<T = any> {
  [index: string]: T
}

export enum VNodeType {
  Element = 'element',
  Text = 'text',
  Unknown = 'unknown'
}

export interface FormItemSchema {
  model?: string
  label?: string
  tag: string
  prop?: Dictionary<string>
}

export interface ButtonSchema {
  type: string
  size: string
  text: string
}

export interface FormSchema {
  formItem: Array<FormItemSchema>
  buttons: Array<ButtonSchema>
}

export interface OptionsSchema extends Dictionary<FormSchema> {
  queryForm: FormSchema
  // displayForm: FormSchema
  // submitForm: FormSchema
}
