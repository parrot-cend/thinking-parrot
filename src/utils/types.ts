export interface Dictionary<T = any> {
  [index: string]: T
}

export type ContextOptions = Dictionary<string>
