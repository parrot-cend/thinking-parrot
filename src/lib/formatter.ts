import prettier from 'prettier'

export function formatByPrettier(str: string): string {
  return prettier.format(str, {
    parser: 'html',
    semi: false,
    bracketSpacing: true,
    htmlWhitespaceSensitivity: 'ignore',
    endOfLine: 'lf'
  })
}
