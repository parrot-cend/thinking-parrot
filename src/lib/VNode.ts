import { formatByPrettier } from './formatter'
import { Dictionary, Node } from './types'

export interface VNodeOptions {
  props?: Dictionary<Node.Prop>
  children?: Array<VNode>
}

export class VNode {
  public tagName: string = ''
  public props: Array<Node.Prop>
  public children: Array<VNode> = []

  constructor(tag: string, props?: Array<Node.Prop> | Node.Prop) {
    this.tagName = tag
    if (props) {
      this.props = Array.isArray(props) ? props : [props]
    } else {
      this.props = []
    }
    return this
  }
  private makeTagStr(prop?: string, children?: string): string {
    return children?.length ? `<${this.tagName}${prop}>${children}</${this.tagName}>` : `<${this.tagName}${prop} />`
  }
  private makePropStr(): string {
    if (this.props.length > 0) {
      return (
        ' ' +
        this.props
          .map(prop => {
            const realKey = /[a-z][A-Z]/.test(prop.key) ? prop.key.replace(/([A-Z])/g, '-$1').toLowerCase() : prop.key
            if (prop.type === Node.PropType.String) {
              return `${realKey}="${prop.value}"`
            } else if (prop.type === Node.PropType.Boolean) {
              return !prop.value ? realKey : `:${realKey}="${prop.value || false}"`
            } else if (prop.type === Node.PropType.Expression) {
              return `:${realKey}="${prop.value}"`
            } else {
              throw new Error(`
                Invalid Prop Type. Node infomation below:\n
                tagName: ${this.tagName}\n
                props: ${JSON.stringify(this.props)}\n
                errorProp: key: ${JSON.stringify(this.props)}
              `)
            }
          })
          .join(' ')
      )
    } else {
      return ''
    }
  }
  private makeChildrenStr(): string {
    return this.children.reduce<string>((all: string, cur: VNode) => {
      if (typeof cur === 'string') {
        return all + cur
      } else {
        return all + cur.toString()
      }
    }, '')
  }
  public toString(inline?: boolean): string {
    const result = this.makeTagStr(this.makePropStr(), this.makeChildrenStr())
    if (inline) {
      return result
    } else {
      return formatByPrettier(result)
    }
  }
  public insertChild(child: VNode | Array<VNode>): this {
    if (Array.isArray(child)) {
      this.children.push(...child)
    } else {
      this.children.push(child)
    }
    return this
  }
  public removeChild(pos: number): this {
    this.children.splice(pos, 1)
    return this
  }
}

export function createNode(tag: string, props?: Array<Node.Prop> | Node.Prop): VNode {
  return new VNode(tag, props)
}
