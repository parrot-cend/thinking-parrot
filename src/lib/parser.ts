import { Node, SourceConfig } from './types'
import { createNode, VNode } from './VNode'

type TemplateCode = string

export default function parser(source: SourceConfig): TemplateCode {
  const makeNodeTree = (conf: Node.Config): VNode => {
    const node = createNode(conf.tag, conf.props)
    if (conf.children?.length) {
      node.insertChild(conf.children.map((c: Node.Config) => makeNodeTree(c)))
    }
    return node
  }
  return source
    .map((conf: Node.Config) => makeNodeTree(conf))
    .map((node: VNode) => node.toString())
    .join('\n')
}
