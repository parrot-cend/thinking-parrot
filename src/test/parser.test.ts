import parser from '../lib/parser'
import { SourceConfig } from '../lib/types'

describe('Parser test', () => {
  it('generate correct template string', () => {
    const input: SourceConfig = [
      {
        tag: 'el-form',
        props: [
          { key: 'inline', type: 'boolean', value: '' },
          { key: 'labelWidth', type: 'string', value: '100px' }
        ],
        children: [
          {
            tag: 'el-form-item',
            props: [
              { key: 'required', type: 'boolean', value: '' },
              { key: 'label', type: 'string', value: '' }
            ],
            children: [
              {
                tag: 'el-input',
                props: [
                  { key: 'type', type: 'string', value: 'text' },
                  { key: 'bind', type: 'expression', value: 'text' },
                  { key: 'type', type: 'string', value: 'text' },
                  { key: 'clearable', type: 'boolean', value: '' }
                ],
                children: []
              }
            ]
          }
        ]
      },
      {
        tag: 'el-form',
        props: [
          { key: 'inline', type: 'boolean', value: 'false' },
          { key: 'labelWidth', type: 'string', value: '100px' }
        ],
        children: []
      }
    ]
    const output = parser(input)
    expect(output).toMatchSnapshot()
  })
  it('throw error when invalid prop type given', () => {
    const input: SourceConfig = [
      {
        tag: 'el-form',
        props: [
          { key: 'inline', type: 'invalidType', value: '' },
          { key: 'labelWidth', type: 'string', value: '100px' }
        ],
        children: []
      }
    ]
    const test = (): any => parser(input)
    expect(test).toThrowError(Error)
  })
})
