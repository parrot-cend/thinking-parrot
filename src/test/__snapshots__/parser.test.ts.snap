// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser test generate correct template string 1`] = `
"<el-form inline label-width=\\"100px\\">
  <el-form-item required label=\\"\\">
    <el-input type=\\"text\\" :bind=\\"text\\" type=\\"text\\" clearable />
  </el-form-item>
</el-form>

<el-form :inline=\\"false\\" label-width=\\"100px\\" />
"
`;
